src/librustc_typeck/check/mod.rs

```diff
@@ -1148,6 +1148,48 @@ fn check_fn<'a, 'gcx, 'tcx>(inherited: &'a Inherited<'a, 'gcx, 'tcx>,
         }
     }
 
+    // Check that a function marked as `#[panic_implementation]` has signature `fn(&PanicInfo) -> !`
```

Comment:
I'm personally more of a fan of generating AST nodes which correspond to the structure we want which avoids reaching into the internals of the typechecker where possible. For example I don't think this disallows something like this:

```rust
#[panic_implementation]
fn foo<T>(a: &PanicInfo) -> ! {
}
```

right?

I was thinking that we'd basically expand `#[panic_implementation]` to `#[lang = "panic_impl"]` where the lang item internally dispatches to `#[panic_implementation]` (automatically typechecking it along the way). That may be slightly more difficult to architect, though, but I'm curious what you think

Reply:
The language items are not type-checked at all at the moment, so there is nothing to hand the type-checking over to.

`panic_fmt` has no benefits over the new language item and IIRC (from when I wrote the instructions) complicates the implementation of this feature unnecessarily.

Reply:
Right yeah, my point is that we *generate* a language item which internally dispatches to the function annotated `#[panic_implementation]`, which by construction will typecheck the panic implementation and not require typechekcing of lang ites.

Reply:
@alexcrichton you mean something like this?

``` rust
// this
// (signature is totally wrong)
#[panic_implementation]
fn foo<T>(pi: &i32) {
    // body
}

// expands into
#[lang = "panic_impl"]
fn __secret_name_maybe(pi: &PanicInfo) -> ! {
    // user input
    fn foo<T>(pi: &i32) {
        // body
    }

    let f: fn(&PanicInfo) -> ! = foo;
    f()
}
```

That would work typechecking wise, but will it report error messages with meaningful spans? (My experience with proc macros says that most errors will wrongly report the whole #[panic_implementation] item as their span)

Reply:
>  I don't think this disallows something like this:
> fn foo<T>(a: &PanicInfo) -> ! {

I can add a check and test to reject that.

Reply:
Sadly in many cases this approach results in kind-of suboptimal errors.

For example


````
fn foo() {
    fn bar<T>(i: i32) -> ! { loop {} }
    let foo: fn(i32) -> ! = bar;
}
````

gives 

````
error[E0282]: type annotations needed
 --> src/main.rs:3:29
  |
3 |     let foo: fn(i32) -> ! = bar;
  |                             ^^^ cannot infer type for `T`
````

but if expanded within a macro it would be something like this instead:

````
error[E0282]: type annotations needed
 --> src/main.rs:3:29
  |
3 |     #[panic_implementation]
  |     fn panic<T>(...) -> ! { loop {} }
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot infer type for `T`
````

which is just… incomprehensible.

That being said, I have a long time wish to eventually implement universal checking of the language item signatures (similar to one we do for intrinsics), which would resolve this issue universally. Thus, I’d recommend going for the simplest complete approach possible for now.